File: plotDataFromWorkspace.py
import ROOT
ROOT.gROOT.SetBatch(True)  # run in batch mode

# Load the workspace from a ROOT file
input_file = ROOT.TFile.Open("datacards_HIG_23_001/cards_2018/HCG/1500/hzz2l2q_13TeV_xs.root")
workspace = input_file.Get("w")

# Get the RooDataSet from the workspace
data_obs = workspace.data("data_obs")

# Define the frame for zz2l2q_mass
frame1 = workspace.var("zz2l2q_mass").frame(ROOT.RooFit.Title("ZZ to 2l2q Mass"))
# Plot data on the frame
data_obs.plotOn(frame1)

# Draw the frame
canvas1 = ROOT.TCanvas("canvas1", "canvas1")
frame1.Draw()
canvas1.SaveAs("zz2l2q_mass_plot.png")

# Define the frame for zz2l2q_mass
frame1 = workspace.var("zz2lJ_mass").frame(ROOT.RooFit.Title("ZZ to 2lJ Mass"))
# Plot data on the frame
data_obs.plotOn(frame1)

# Draw the frame
canvas12 = ROOT.TCanvas("canvas12", "canvas12")
frame1.Draw()
canvas12.SaveAs("zz2lJ_mass_plot.png")

# Similarly you can define frames for other variables and plot them
frame2 = workspace.var("Dspin0").frame(ROOT.RooFit.Title("Spin Discriminant"))
data_obs.plotOn(frame2)

canvas2 = ROOT.TCanvas("canvas2", "canvas2")
frame2.Draw()
canvas2.SaveAs("Dspin0_plot.png")

# Define the frame for zz2l2q_mass
frame1 = workspace.var("CMS_channel").frame(ROOT.RooFit.Title("ZZ to 2lJ Mass"))
# Plot data on the frame
data_obs.plotOn(frame1)

# Draw the frame
canvas3 = ROOT.TCanvas("canvas3", "canvas3")
frame1.Draw()
canvas3.SaveAs("CMS_channel_plot.png")

# Don't forget to close the ROOT file
input_file.Close()

File: plotDataFromWorkspace.cpp
#include "RooWorkspace.h"
#include "RooPlot.h"
#include "RooRealVar.h"
#include "RooDataSet.h"
#include "TCanvas.h"
#include "TFile.h"

void plotDataFromWorkspace() {

  // Open the ROOT file containing the workspace
  TFile* f = new TFile("datacards_HIG_23_001/cards_2018/HCG/1500/hzz2l2q_13TeV_xs.root");

  // Retrieve workspace from ROOT file
  RooWorkspace* ws = (RooWorkspace*) f->Get("w");

  // Retrieve the dataset from the workspace
  RooDataSet* data_obs = (RooDataSet*) ws->data("data_obs");

  // Create a canvas for the plots
  TCanvas* c1 = new TCanvas("c1","c1",800,600);
  c1->Divide(2,2);  // Create a 2x2 grid on the canvas

  // Plotting CMS_channel
  c1->cd(1);
  RooRealVar* CMS_channel = ws->var("CMS_channel");
  RooPlot* frame1 = CMS_channel->frame();
  data_obs->plotOn(frame1);
  frame1->Draw();

  // Plotting zz2l2q_mass
  c1->cd(2);
  RooRealVar* zz2l2q_mass = ws->var("zz2l2q_mass");
  RooPlot* frame2 = zz2l2q_mass->frame();
  data_obs->plotOn(frame2);
  frame2->Draw();

  // Plotting Dspin0
  c1->cd(3);
  RooRealVar* Dspin0 = ws->var("Dspin0");
  RooPlot* frame3 = Dspin0->frame();
  data_obs->plotOn(frame3);
  frame3->Draw();

  // Plotting zz2lJ_mass
  c1->cd(4);
  RooRealVar* zz2lJ_mass = ws->var("zz2lJ_mass");
  RooPlot* frame4 = zz2lJ_mass->frame();
  data_obs->plotOn(frame4);
  frame4->Draw();

  c1->Update();

  // Save the canvas to a file
  c1->SaveAs("data_obs_plots.png");

}

int main() {
  plotDataFromWorkspace();
  return 0;
}
